Experimental: Add a local MCP server with the ability to translate in parallel

hassoncs opened the pull request:
## Feature: Local Model Context Protocol (MCP) Integration for faster dev i18n

This PR introduces a local Model Context Protocol (MCP) server to significantly enhance our internationalization (i18n) workflow, particularly for managing translation strings.

**Why is this cool?**

By delegating translation tasks to a local MCP server, we can now batch process translations in parallel, leading to a substantial increase in efficiency.

**Performance Highlights:**

Our speed tests demonstrate the power of this new approach:

*   **Adding new strings:** Adding three new strings across 17 locales manually took over 2 minutes. With the MCP tool, the same task was completed in just 14 seconds – **nearly 90% faster!**
*   **Removing strings:** Manually removing the same three strings from all locales took over 2 minutes. Using the MCP tool, the removal was instantaneous – **99% faster!**

This integration streamlines the translation management process, saving valuable development time.

**Technical Details:**

*   Implemented a local MCP Server that automatically starts up as a task in the background.
*   Enabled translation of specific keys, translating missing keys, moving keys, and listing locales via the MCP server.
*   Created a VSCode task for starting the MCP Translation Server.

This is a significant step towards a more efficient and scalable i18n workflow and towards parallel agent tasks!

drakonen commented:
Very cool, hope to see it in a good state soon!

seuros commented:
Appreciate the wild spirit here, like the world revolves around English-to-X translations or something.

That said, I’ve got a few concerns:

1. Scope Creep
Embedding a specialized translation MCP server directly in the core extension feels off. The extension’s purpose should stay broad and context-agnostic. This introduces assumptions that devs are always working on web apps with i18n needs, which isn’t always true. Some of my projects don’t even involve text,  they don’t deal with humans.

2. LLM Confusion
Having this run silently in the background means the extension starts piping translation-related context into the agent, even when it’s irrelevant. That introduces unnecessary noise and risks messing with prompt clarity when the agent is supposed to be doing literally anything else.

3. Tech Stack Choices
I also noticed this routes through OpenRouter. That’s fine, but feels overkill for a utility that could just use something simpler and cheaper, like Google Translate, which is also surprisingly good with low-context phrases.


This translation module should not be part of the core extension. Like other MCP servers, it deserves its own repo and should live as a standalone opt-in tool for those who need it.

hassoncs commented:
Hey @seuros, 

I definitely could have explained things better in my description- your concerns are 100% valid if this translation MCP was something that shipped as part of the core extension. But just to clarify: none of this runs inside the packaged extension (if it is I made a mistake!). It’s purely a local dev utility, meant to speed up our own workflows while working on the extension. So the assumptions it makes (about i18n, web apps, text, etc.) only apply during our dev, not for end users.

It won’t run unless you’re actively developing the extension and have the repo pulled down—so no agents in other environments are ever piping this translation context in. The main motivation was just cutting down the tedium of i18n dev work locally.

Re: OpenRouter—I totally agree it might be overkill for a utility like this. I just followed the precedent set by [Roo’s translate mode rules here](https://github.com/RooCodeInc/Roo-Code/tree/main/.roo/rules-translate).

Longer term, I agree this could probably be spun off into a standalone, opt-in repo (like other MCP tools). This was all vibe-coded to solve a problem in the moment, and if it proves useful, I think we can factor it out cleanly.

Appreciate you flagging this—definitely want to avoid anything that creeps into core or muddies up the architecture.

seuros commented:
Oh, sorry , i was looking a the codebase then checked in the PR assuming it was another tool like read_file. 

Seems like i did exactly what i instruct the LLM to not do "Assume architecture".  `Do what i say, not what i will do :D`

hassoncs committed:
wip

hassoncs committed:
'chore: WIP'

hassoncs committed:
🔧 chore(gitignore): remove unused ignore entries

- remove .mcp-server-running and trailing newline for cleanup

♻️ refactor(mcp): update server command execution

- change command to use bash for better path handling
- simplify args to execute directly without separate cwd

📝 docs(README): update setup and workflow instructions

- add tsx as a required dev dependency
- remove build and start steps for direct execution

♻️ refactor(package): modify main entry point and scripts

- change main from build/index.js to src/index.ts
- remove build and start scripts for direct execution via tsx

hassoncs committed:
Add remove_i18n_keys tool and improve environment variable handling

hassoncs committed:
🔧 chore(env): add OpenRouter API key to sample

- Update .env.sample to include a placeholder for OpenRouter API key

♻️ refactor(mcp): enhance server configuration and logging

- Replace translation-mcp-server with repo-mcp-server in config
- Add nodemon for auto-restart on code changes
- Update environment variable loading and logging
- Enhance API key validation and logging

🛠️ chore(deps): update package-lock for node version compatibility

- Adjust node engine version to allow >=18.20.8
- Remove unnecessary dependencies from package-lock

✨ feat(i18n): validate OpenRouter API key presence

- Add checks for OpenRouter API key in translation tool
- Log error when API key is missing in context

♻️ refactor(translation): improve API key validation and logging

- Simplify validation logic for API key
- Add masked logging for API key to improve security visibility

hassoncs committed:
♻️ refactor(index): improve project root path resolution

- use __dirname to reliably find the project root by resolving two directories up

hassoncs committed:
Fix the mcp server not starting

hassoncs committed:
♻️ refactor(server): simplify code and improve error handling

- remove redundant comments and console logs for clearer code
- simplify environment variable loading and error handling
- refactor function structures for consistency and readability

♻️ refactor(i18n): streamline i18n tool handlers

- clean up i18n tool handler code by removing unnecessary comments
- simplify key expansion and validation logic
- improve error handling and logging for translation tasks

♻️ refactor(utils): optimize utility functions

- remove excessive comments and inline documentation
- refine function logic for handling JSON and locale operations
- consolidate error handling in utility functions for consistency

hassoncs committed:
✨ feat(i18n): add indentation detection for translation files

- introduce indentation detection from existing English files to maintain consistency
- apply detected indentation to newly created or updated translation files
- handle errors in file reading for indentation detection gracefully

hassoncs commented on package.json:
```diff
@@ -11,7 +11,7 @@
 	"engines": {
 		"vscode": "^1.84.0",
-		"node": "20.18.1"
+		"node": ">=18.20.8"
```
I think it was because of a build warning... I reverted this change– I don't think it's needed!

hassoncs committed:
'chore: WIP'

EamonNerbonne commented on repo-mcp-server/src/utils/locale-utils.ts:
```diff
@@ -20,44 +20,9 @@ export async function getI18nNamespaces(
-
-	const baseCode = locale.split("-")[0].toLowerCase()
-	return languageMap[baseCode] || baseCode
+// Function to get the locale code in a format suitable for translation
+export function getLocaleForTranslation(locale: string): string {
+	// Just return the locale code itself, which is what most translation APIs expect
+	// This avoids maintaining a map of locale codes to language names
+	return locale
```
inline, since no-op?

hassoncs replied:
removed!

EamonNerbonne commented on repo-mcp-server/README.md:
```diff
@@ -0,0 +1,73 @@
+The server looks for an `.env.local` file in the parent directory for configuration variables:
+
+- `OPENROUTER_API_KEY` - API key for OpenRouter translation service
+- `DEFAULT_MODEL` - Default model to use for translation (defaults to "anthropic/claude-3.7-sonnet")
```
As discussed yesterday, edit readme so it's clear this is just a dev-tool, not something we ship, and also that it's purely a performance optimization; we can just delete it when maintenance burden outweighs the time spent waiting for the LLM.

EamonNerbonne commented on repo-mcp-server/src/tools/README.md:
```diff
@@ -0,0 +1,166 @@
+4. **Consistent Error Handling**: Follow the established error handling pattern for consistent behavior.
+5. **Logging**: Use console.error for logging to help with debugging.
+6. **Modular Design**: Keep each tool focused on a specific task for better maintainability.
+7. **Testing**: Consider adding unit tests for your tools to ensure they work as expected.
```
Delete file?

hassoncs committed:
📝 docs(README): update and simplify documentation

- remove detailed structure and development sections
- emphasize the tool's purpose and usage
- update configuration requirements

🔥 chore(i18n): remove unused i18n tools README

- delete i18n tools documentation as it's no longer needed

♻️ refactor(i18n): enhance indentation handling for i18n tools

- add detection and application of source file indentation
- improve logging for indentation detection process
- ensure consistent formatting across translated files

hassoncs committed:
♻️ refactor(server): streamline tool handling

- replace getAllTools and getToolByName functions with direct i18nTools usage
- remove redundant tools index file to simplify code structure

hassoncs committed:
📦 build(dependencies): update comment-json to version 4.2.5

- update comment-json from version 4.2.3 to 4.2.5 in package.json
- add new dependencies in package-lock.json

♻️ refactor(i18n-tools): simplify indentation detection and usage

- refactor indentation detection to return a string directly
- replace usage of IndentationSettings interface with string
- adjust logging and JSON stringification code accordingly

hassoncs committed:
♻️ refactor(i18n): use comment-json for preserving formatting

- integrate comment-json library to maintain JSON formatting across i18n tools
- replace JSON.stringify with commentJson.stringify to preserve comments and formatting
- update indentation detection logic to ensure correct usage of tabs or spaces
